homie/src/lib.rs

```diff
@@ -282,13 +332,68 @@ impl HomieDevice {
+    ) -> impl Future<Output = Result<(), Box<dyn Error + Send + Sync>>> {
+        let device_base = format!("{}/", self.publisher.device_base);
+        let (incoming_tx, incoming_rx) = async_channel::unbounded();
+
+        let mqtt_task = task::spawn(async move {
             loop {
                 let (incoming, outgoing) = event_loop.poll().await?;
                 log::trace!("Incoming = {:?}, Outgoing = {:?}", incoming, outgoing);
+
+                if let Some(incoming) = incoming {
+                    incoming_tx.send(incoming).await?;
+                }
             }
-        })
+        });
+
+        let publisher = self.publisher.clone();
+        let incoming_task: JoinHandle<Result<(), Box<dyn Error + Send + Sync>>> =
+            task::spawn(async move {
+                loop {
+                    match incoming_rx.recv().await? {
+                        Incoming::Publish(publish) => {
+                            if let Some(rest) = publish.topic.strip_prefix(&device_base) {
+                                if let ([node_id, property_id, "set"], Ok(payload)) = (
+                                    rest.split("/").collect::<Vec<&str>>().as_slice(),
+                                    str::from_utf8(&publish.payload),
+                                ) {
+                                    log::trace!(
+                                        "set node {:?} property {:?} to {:?}",
+                                        node_id,
+                                        property_id,
+                                        payload
+                                    );
+                                    if let Some(callback) = update_callback.as_mut() {
+                                        if callback(
+                                            node_id.to_string(),
+                                            property_id.to_string(),
+                                            payload.to_string(),
+                                        )
+                                        .await
+                                        {
+                                            publisher
+                                                .publish_retained(
+                                                    &format!("{}/{}", node_id, property_id),
+                                                    payload,
+                                                )
+                                                .await?;
+                                        }
+                                    }
+                                }
+                            } else {
+                                log::warn!("Unexpected publish: {:?}", publish);
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+            });
+        try_join_handles(mqtt_task, incoming_task).map(|r| r.map(|((), ())| ()))
```

Comment:
I feel like `try_join_handles()` is a bit too generic. This type signature is common enough that we probably want a helper called `try_join_unit_handles()` or something.

Reply:
Done.